zephyr/lvgl.c

```diff
@@ -0,0 +1,358 @@
+		    cap.current_orientation == DISPLAY_ORIENTATION_ROTATED_180) {
+			prev.point.x = cap.x_resolution - prev.point.x;
+		} else {
+			prev.point.x = cap.y_resolution - prev.point.x;
+		}
+	}
+
+	if (IS_ENABLED(CONFIG_LV_Z_POINTER_KSCAN_INVERT_Y)) {
+		if (cap.current_orientation == DISPLAY_ORIENTATION_NORMAL ||
+		    cap.current_orientation == DISPLAY_ORIENTATION_ROTATED_180) {
+			prev.point.y = cap.y_resolution - prev.point.y;
+		} else {
+			prev.point.y = cap.x_resolution - prev.point.y;
+		}
+	}
+
+	/* rotate touch point to match display rotation */
+	if (cap.current_orientation == DISPLAY_ORIENTATION_ROTATED_90) {
+		lv_coord_t x;
+
+		x = prev.point.x;
+		prev.point.x = prev.point.y;
+		prev.point.y = cap.y_resolution - x;
+	} else if (cap.current_orientation == DISPLAY_ORIENTATION_ROTATED_180) {
+		prev.point.x = cap.x_resolution - prev.point.x;
+		prev.point.y = cap.y_resolution - prev.point.y;
+	} else if (cap.current_orientation == DISPLAY_ORIENTATION_ROTATED_270) {
+		lv_coord_t x;
+
+		x = prev.point.x;
+		prev.point.x = cap.x_resolution - prev.point.y;
+		prev.point.y = x;
+	}
+
+set_and_release:
+	*data = prev;
+
+	k_msgq_num_used_get(&kscan_msgq);
+}
+
+static int lvgl_pointer_kscan_init(void)
+{
+	const struct device *kscan_dev =
+		device_get_binding(CONFIG_LV_Z_POINTER_KSCAN_DEV_NAME);
+
+	if (kscan_dev == NULL) {
+		LOG_ERR("Keyboard scan device not found.");
+		return -ENODEV;
+	}
+
+	if (kscan_config(kscan_dev, lvgl_pointer_kscan_callback) < 0) {
+		LOG_ERR("Could not configure keyboard scan device.");
+		return -ENODEV;
+	}
+
+	lv_indev_drv_init(&indev_drv);
+	indev_drv.type = LV_INDEV_TYPE_POINTER;
+	indev_drv.read_cb = lvgl_pointer_kscan_read;
+
+	if (lv_indev_drv_register(&indev_drv) == NULL) {
+		LOG_ERR("Failed to register input device.");
+		return -EPERM;
+	}
+
+	kscan_enable_callback(kscan_dev);
+
+	return 0;
+}
+#endif /* CONFIG_LV_Z_POINTER_KSCAN */
+
+static int lvgl_init(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	const struct device *display_dev =
+		device_get_binding(CONFIG_LV_Z_DISPLAY_DEV_NAME);
+	int err = 0;
+
+	if (display_dev == NULL) {
+		LOG_ERR("Display device not found.");
+		return -ENODEV;
+	}
+
+#if CONFIG_LV_LOG_LEVEL != 0
+	lv_log_register_print_cb(lvgl_log);
+#endif
+
+	lv_init();
+
+#ifdef CONFIG_LV_Z_USE_FILESYSTEM
+	lvgl_fs_init();
+#endif
+
+	lv_disp_drv_init(&disp_drv);
+	disp_drv.user_data = (void *) display_dev;
+
+	err = lvgl_allocate_rendering_buffers(&disp_drv);
+	if (err != 0) {
+		return err;
+	}
+
+	if (set_lvgl_rendering_cb(&disp_drv) != 0) {
+		LOG_ERR("Display not supported.");
+		return -ENOTSUP;
+	}
+
+	if (lv_disp_drv_register(&disp_drv) == NULL) {
+		LOG_ERR("Failed to register display device.");
+		return -EPERM;
+	}
+
+#ifdef CONFIG_LV_Z_POINTER_KSCAN
+	lvgl_pointer_kscan_init();
+#endif /* CONFIG_LV_Z_POINTER_KSCAN */
+
+	return 0;
+}
+
+SYS_INIT(lvgl_init, APPLICATION, CONFIG_APPLICATION_INIT_PRIORITY);
```

Comment:
Since so much is being changed, I think the options LV_Z_DISPLAY_DEV_NAME,  LV_Z_HOR_RES_MAX,  LV_Z_VER_RES_MAX and this SYS_INIT should also be removed right away, lvgl_init() reworked and it should be left to the user to initialize everything. Otherwise there will be dependencies that are hard to maintain.

Reply:
I'd honestly just move on with LVGL upgrade for now. Display drivers and LVGL glue code need a deep rework, not only this change alone, IMO.